docs/proposals/memory-management.md

```diff
@@ -0,0 +1,185 @@
+
+6. Detect OOM kills on containers: Docker tracks OOM kills on its containers.
+Kubelet exposes this information via [ContainerStatus.State](../../pkg/api/v1/types.go#L1033), where the state is set to `ContainerStateTerminated` and `Reason` is set to `OOM Killed`.
+Kubelet has a built-in exponential backoff mechanism that prevents OOMing containers from entering a crash loop and bringing down the node.
+
+
+#### What are some of the known limitations?
+
+1. The scheduler does not take into account the total usage on the node. Hence, memory can be easily over-committed. Whenever a job is scheduled without resource requests, the node becomes over-committed.
+
+2. Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+
+3. A container in a pod with restart policy set to `Always` will be restarted continuously even if it OOMs. This is alleviated with the crash loop backoff feature. To improve node stability, that pod should be evicted and run on a different node.
+
+4. Other than the Guaranteed class, containers in other classes don't have a memory limit.
+A `Best-effort` container can use a lot of memory and result in high tail latencies for `Guaranteed` and `Burstable` class containers.
+Memory allocations will hang until the kernel can free up memory, by identifying and killing a process in a container in a lower class.
+
+5. Kubelet does not act on OOM kills.
+A container can choose to continue running when one or more processes in the container were killed by the kernel OOM killer.
+Because of this, the kubelet cannot deterministically associate container death with OOMs.
+This affects higher level systems like auto-scaler which requires reliable OOM signals to update memory limits.
+
+6. In the event of an OOM kill due to system memory pressure, the kubelet cannot identify the container of the killed process.
+
+7. All burstable containers that exceed their request are equally likely to be killed. Ideally, the job that exceeds the most should be killed.
+
+8. Memory usage of system daemons is unrestricted. This results in varying amount of memory availability on the node.
+
+
+##### How to address these limitations?
+
+We can improve memory management by making use of [kernel memcg features](https://www.kernel.org/doc/Documentation/cgroups/memory.txt), and by modifying the current system behavior. Some of these improvements are required in the short-term. The rest are either complex or require significant changes to the system.
+
+###### What can be done right away?
+
+1. Cleanup the entire container whenever a OOM kill is detected in a container.
+Users should be able to override the default behavior for containers that have an in-built process manager that can handle OOM kills.
+
+2. Set memory soft limits on Burstable containers to their request.
+This should result in "Burstable" containers being pushed to their requested memory limit under system memory pressure.
```

Comment:
This makes sense and seems like a good thing to do in the near term.  I am not sure how we balance this feature with the idea of differentiating containers in the same class based on how much resource they have consumed beyond their requested value.  It seems mutually exclusive.  In practice, could this not mean that any container using more than its requested usage could be OOMKilled proactively and we would have no means of comparison?

As for benefit, maybe just take the following:

"When the system detects memory contention or low memory, the container is forced to restrict its consumption to its soft limits.   If lowering the memory usage to the soft limit does not solve the contention, the container are pushed back as much as possible to make sure that one control group does not starve the others of memory. Note that soft limits take effect over a long period of time, since they involve reclaiming memory for balancing between memory containers."

We should create an issue for this and link.